crates/build/src/lib.rs

```diff
@@ -927,10 +807,371 @@ pub fn execute(args: ExecuteArgs) -> Result<BuildResult> {
+    tracing::debug!(
+        "Fingerprint before build: {:?}, after build: {:?}",
+        pre_fingerprint,
+        post_fingerprint
+    );
+
+    let dest_code_path = crate_metadata.dest_code.clone();
+
+    if pre_fingerprint == Some(post_fingerprint) && crate_metadata.dest_code.exists() {
+        tracing::info!(
+            "No changes in the original wasm at {}, fingerprint {:?}. \
+                Skipping Wasm optimization and metadata generation.",
+            crate_metadata.original_code.display(),
+            pre_fingerprint
+        );
+        return Ok((None, build_info, dest_code_path, build_steps))
+    }
+
+    maybe_lint(
+        &mut build_steps,
+        *build_artifact,
+        *lint,
+        crate_metadata,
+        verbosity,
+    )?;
+
+    maybe_println!(
+        verbosity,
+        " {} {}",
+        format!("{build_steps}").bold(),
+        "Post processing code".bright_green().bold()
+    );
+    build_steps.increment_current();
+
+    // remove build artifacts so we don't have anything stale lingering around
+    for t in Target::iter() {
+        fs::remove_file(crate_metadata.dest_code.with_extension(t.dest_extension())).ok();
+    }
+
+    let original_size =
+        fs::metadata(&crate_metadata.original_code)?.len() as f64 / 1000.0;
+
+    match target {
+        Target::Wasm => {
+            post_process_wasm(
+                crate_metadata,
+                *skip_wasm_validation,
+                verbosity,
+                *max_memory_pages,
+            )?;
+            let handler = WasmOptHandler::new(*optimization_passes, *keep_debug_symbols)?;
+            handler.optimize(
+                &crate_metadata.dest_code,
+                &crate_metadata.contract_artifact_name,
+            )?;
+        }
+        Target::RiscV => {
+            fs::copy(&crate_metadata.original_code, &crate_metadata.dest_code)?;
+        }
+    }
+
+    let optimized_size = fs::metadata(&dest_code_path)?.len() as f64 / 1000.0;
+
+    let optimization_result = OptimizationResult {
+        original_size,
+        optimized_size,
+    };
+
+    Ok((
+        Some(optimization_result),
+        build_info,
+        crate_metadata.dest_code.clone(),
+        build_steps,
+    ))
+}
+
+pub fn maybe_lint(
+    steps: &mut BuildSteps,
+    build_artifact: BuildArtifacts,
+    lint: bool,
+    crate_metadata: &CrateMetadata,
+    verbosity: &Verbosity,
+) -> Result<()> {
+    let total_steps = build_artifact.steps();
+    if lint {
+        steps.set_total_steps(total_steps + 1);
+        maybe_println!(
+            verbosity,
+            " {} {}",
+            format!("{steps}").bold(),
+            "Checking ink! linting rules".bright_green().bold()
+        );
+        steps.increment_current();
+        exec_cargo_dylint(crate_metadata, *verbosity)?;
+        Ok(())
+    } else {
+        steps.set_total_steps(total_steps);
+        Ok(())
+    }
+}
+
+pub fn docker_build(args: ExecuteArgs) -> Result<BuildResult> {
```

Comment:
I reckon move this into its own `docker.rs` file, this file is getting way too big now.